app/src/androidTest/java/com/github/se/eduverse/ui/search/UserProfileScreenTest.kt

```diff
@@ -240,4 +252,82 @@ class UserProfileScreenTest {
     composeTestRule.onNodeWithTag("publication_item_pub1").assertDoesNotExist()
     composeTestRule.onNodeWithTag("publication_item_fav1").assertExists()
   }
+
+  @Test
+  fun verifyFollowButtonStates() {
+    val testProfile =
+        Profile(id = "other_user_id", username = "TestUser", isFollowedByCurrentUser = false)
+
+    fakeViewModel.setState(ProfileUiState.Success(testProfile))
+    fakeViewModel.setFollowActionState(FollowActionState.Idle)
+
+    composeTestRule.setContent {
+      UserProfileScreen(
+          navigationActions = fakeNavigationActions,
+          viewModel = fakeViewModel,
+          userId = "other_user_id", // Profile we're viewing
+          currentUserId = "current_user_id" // Logged in user
+          )
+    }
+
+    // Debug prints to help identify the issue
+    println("Profile ID: other_user_id")
+    println("Current User ID: current_user_id")
+
+    composeTestRule
+        .onNodeWithTag("follow_button")
+        .assertExists()
+        .assertTextContains("Follow")
+        .assertIsEnabled()
+
+    fakeViewModel.setFollowActionState(FollowActionState.Loading)
+    composeTestRule.onNodeWithTag("follow_button").assertExists().assertIsNotEnabled()
+
+    fakeViewModel.setFollowActionState(FollowActionState.Error("Follow failed"))
+    composeTestRule.onNodeWithText("Follow failed").assertExists()
+
+    val updatedProfile = testProfile.copy(isFollowedByCurrentUser = true)
+    fakeViewModel.setState(ProfileUiState.Success(updatedProfile))
+    fakeViewModel.setFollowActionState(FollowActionState.Success)
+    composeTestRule.onNodeWithTag("follow_button").assertExists().assertTextContains("Unfollow")
+  }
+
+  @Test
+  fun whenFollowButtonClicked_showsLoadingAndUpdatesState() {
+    val testProfile =
+        Profile(id = "other_user_id", username = "TestUser", isFollowedByCurrentUser = false)
+
+    fakeViewModel.setState(ProfileUiState.Success(testProfile))
+    fakeViewModel.setFollowActionState(FollowActionState.Idle)
+
+    composeTestRule.setContent {
+      UserProfileScreen(
+          navigationActions = fakeNavigationActions,
+          viewModel = fakeViewModel,
+          userId = "other_user_id", // Profile we're viewing
+          currentUserId = "current_user_id" // Logged in user
+          )
+    }
+
+    // Debug prints to help identify the issue
+    println("Profile ID: other_user_id")
+    println("Current User ID: current_user_id")
+
+    composeTestRule
+        .onNodeWithTag("follow_button")
+        .assertExists()
+        .assertTextContains("Follow")
+        .performClick()
+
+    fakeViewModel.setFollowActionState(FollowActionState.Loading)
+    composeTestRule.onNodeWithTag("follow_button").assertExists().assertIsNotEnabled()
+
+    fakeViewModel.setFollowActionState(FollowActionState.Error("Failed to follow"))
+    composeTestRule.onNodeWithText("Failed to follow").assertExists()
+  }
+
+  @After
+  fun tearDown() {
+    unmockkAll()
+  }
```

Comment:
The UserProfileScreenTest class is well structured, efficiently covering the main UI states, such as Idle, Loading, and Error. The use of testTags and the optimistic update check are strong points. To improve readability and maintainability, it would be useful to modularize the repeated initialization and remove the debugging printlns. Overall, the class provides good coverage, with some minor tweaks to improve clarity.